src/typesystem.rs

```diff
@@ -559,9 +559,60 @@ fn get_rank(type_information: &DataTypeInformation) -> u32 {
     }
 }
 
-fn is_same_type_nature(ltype: &DataTypeInformation, rtype: &DataTypeInformation) -> bool {
-    (ltype.is_int() && ltype.is_int() == rtype.is_int())
-        || (ltype.is_float() && ltype.is_float() == rtype.is_float())
+pub fn is_same_type_nature(
+    ltype: &DataTypeInformation,
+    rtype: &DataTypeInformation,
+    index: &Index,
+) -> bool {
+    let ltype = find_inner_type_recursive(ltype, index);
+    let rtype = find_inner_type_recursive(rtype, index);
+    match ltype {
+        DataTypeInformation::Struct { name: lname, .. } => {
+            matches!(rtype, DataTypeInformation::Struct { name, ..} if name == lname)
+        }
+        DataTypeInformation::Array { name: lname, .. } => {
+            matches!(rtype, DataTypeInformation::Array { name, ..} if name == lname)
+        }
+        DataTypeInformation::Pointer {
+            inner_type_name: lname,
+            ..
+        } => {
+            matches!(rtype, DataTypeInformation::Pointer{ inner_type_name, ..} if inner_type_name == lname)
+        }
+        DataTypeInformation::Integer { .. } => matches!(rtype, DataTypeInformation::Integer { .. }),
+        DataTypeInformation::Enum { .. } => matches!(rtype, DataTypeInformation::Enum { .. }),
+        DataTypeInformation::Float { .. } => matches!(rtype, DataTypeInformation::Float { .. }),
+        DataTypeInformation::String {
+            size: lsize,
+            encoding: lencoding,
+        } => {
+            matches!(rtype,DataTypeInformation::String { size, encoding } if lsize == size && lencoding == encoding)
+        }
+        DataTypeInformation::Void => matches!(rtype, DataTypeInformation::Void),
+        _ => false,
+    }
+}
+
+fn find_inner_type_recursive<'t>(
```

Comment:
can we rename this to find_intrinsic_type and move it to the index

doc:
/// returns the intrinsic (= built in) type represented by the given type-information.
/// this will return the built-in types behind alias and range-types